### ISSUE TITLE
New option to enable/disable linking to the static C runtime library (Windows, MSVC compiler)

### ISSUE THREAD

**Issue by pedrolcl** [FluidSynth/fluidsynth#840]
### Is your feature request related to a problem?
Right now, the cmake build system forces linking all targets to the static C runtime library when building with MSVC. Applications linking dynamically to the CRT and the fluidsynth library may experience problems like the described in this ticket:
https://github.com/rncbc/qsynth/issues/39

This is a well known problem, as explained here:
https://docs.microsoft.com/en-us/answers/questions/240332/assert-when-using-mtd-dynamic-library.html

### Describe the solution you'd like
A new cmake option "enable linking the static C runtime library when compiling with MSVC", enabled by default.
When disabled, all targets would be dynamically linked to the C runtime DLL.

### Describe alternatives you've considered
This has impacted qsynth, a fluidsynth GUI, which is a Qt application. The alternative of rebuilding each version of the Qt libraries was discarded because it is very time consuming.

**Comment by jjceresa** [comment]
Good point Pedro !. This will avoid a lot of problems inferred  by the static C runtime library.  Also when using the static runtime library we got a dll with a bigger size.

**Comment by jjceresa** [comment]
>A new cmake option "enable linking the static C runtime library when compiling with MSVC", enabled by default.
When disabled, all targets would be dynamically linked to the C runtime DLL.

I suggest that the option should be disabled by default.

**Comment by pedrolcl** [comment]
> I suggest that the option should be disabled by default.

I didn't want to change the current policy, that seems to be needed by some CI environments, but yes: I want this option to disable it in my own boxes, and allow qsynth builds that do not bomb! ;-)

**Comment by jjceresa** [comment]
>I didn't want to change the current policy, that seems to be needed by some CI environments

Linking with a static C runtime library was introduced only for the production of binary dll for Windows XP by the CI builds. In the case of the CI build could be configured, it would be preferable to have an option disabled by default.
@derselbst, please any opinion about how to configure the CI ?.

**Comment by derselbst** [comment]
In principle, I'm open to make this configurable. However, I think the motivation for this step wrong: Pedro, according to the link you've shared, it seems to me that qsynth tries to `free()` memory that fluidsynth has `malloc()`ed. If this is the case, qsynth should use `fluid_free()` to free the memory, rather than `free()`. @pedrolcl Can you please provide a call stack to confirm this?

I predict that qsynth crashes exactly here: https://github.com/rncbc/qsynth/blob/fc350b25bc77911a381153814dfc82d7fb5baa1c/src/qsynthSetupForm.cpp#L171

If so, the proper solution should be to call `fluid_free()` instead of `free()`.

**Comment by pedrolcl** [comment]
You are correct. That line is the cause of the problem, and the call stack confirms it:

~~~
 	ucrtbased.dll!0130bca1()	
 	ucrtbased.dll!
 	[Código externo]	
>	qsynth.exe!qsynth_settings_foreach(void * pvData, const char * pszName, int iType) Línea 171	C++
 	libfluidsynth-3.dll!fluid_settings_foreach(_fluid_hashtable_t * settings, void * data, void(*)(void *, const char *, int) func) Línea 1956	C
 	qsynth.exe!qsynthSetupForm::setup(qsynthOptions * pOptions, qsynthEngine * pEngine, bool bNew) Línea 417	C++
 	qsynth.exe!qsynthMainForm::setupEngineTab(qsynthEngine * pEngine, int iTab) Línea 1633	C++
 	qsynth.exe!qsynthMainForm::showSetupForm() Línea 1731	C++
 	qsynth.exe!qsynthMainForm::qt_static_metacall(QObject * _o, QMetaObject::Call _c, int _id, void * * _a) Línea 164	C++
 	[Código externo]	
 	qsynth.exe!main(int argc, char * * argv) Línea 550	C++
 	qsynth.exe!WinMain(HINSTANCE__ * __formal, HINSTANCE__ * __formal, char * __formal, int __formal) Línea 97	C++
 	[Código externo]	
~~~

I will propose a fix to Rui @rncbc with this change, but it would not be a complete solution for this kind of problem. First, `fluid_free` exists only [since 2019](https://github.com/FluidSynth/fluidsynth/pull/562) in fluidsynth v2.0.7, while qsynth [introduced](https://github.com/rncbc/qsynth/commit/063b1051a33371638a5554e1eef460e030208c11) that line in 2018. Second, qsynth still supports fluidsynth v1.x, and third: `fluid_free` documentation says that the motivation for this function was:
>Especially useful when using fluidsynth with programming languages that do not provide malloc() and free().

Can you guarantee that there are no more `malloc`/`free` conflicts between qsynth and fluidsynth that will trigger this assert in Microsoft's CRT ? The windows distribution of qsynth has been always compiled with mingw, so this problem was never exposed to the users, but do you feel comfortable with the idea of bombing users with this kind of errors instead of avoiding them ?

Conclusion: my proposal stands. I would like to choose between linking fluidsynth with static or dynamic CRT. I would use the static version for testing purposes, but always distribute the dynamic one.

**Comment by derselbst** [comment]
> First, fluid_free exists only since 2019 in fluidsynth v2.0.7, while qsynth introduced that line in 2018. Second, qsynth still supports fluidsynth v1.x,

Which means that qsynth also suffers from various other bugs and crashes that you buy when using an outdated fluidsynth version. Not really an argument for me.

> and third: fluid_free documentation says that the motivation for this function was:

I am aware that the motivation of introducing `fluid_free` was a little different back then. But this doesn't make it less right. I'll update the API docs of `fluid_free` to mention this issue.

> Can you guarantee that there are no more malloc/free conflicts between qsynth and fluidsynth that will trigger this assert in Microsoft's CRT

I don't think it's my job to give those kind of "bug-free-guarantee". I can, however, tell you that `fluid_settings_dupstr()` is currently the only public function that passes a malloced object to the client, who is supposed to free it himself. And the API doc of `fluid_settings_dupstr` correctly mentions this.

> but do you feel comfortable with the idea of bombing users with this kind of errors instead of avoiding them ?

"bombing users" because of this single programming error is quite an impertinent exaggeration. The statically linked CRT made a programming error in qsynth obvious, which IMO is a good thing because it solves the root cause. Switching to dynamic CRT only cures the symptoms. This is not my idea of software development, sorry.

If it makes you happy, I'll take your PR. It would just be nice to have a reason for making things more configurable. I switched to the static CRT almost 3 years ago. And I must admit that I cannot really comprehend why this is suddenly a problem.

**Comment by mawe42** [comment]
If I understand correctly, the problem exists since 2018 with QSynth using FluidSynth >= 2.0.0rc1 compiled with MSVC (assuming QSynth always had this `free` call). So introducing the static/dynamic CRT switch or the fix to QSynth will have no impact on already compiled versions of FluidSynth and/or QSynth. 

It's great that QSynth is fixing this to use `fluid_free()`. And I think it would be nice to give users the choice between static/dynamic CRT for MSVC builds of FluidSynth. Reducing the DLL size is

**Comment by mawe42** [comment]
(Sorry for accidentally closing this, slipped with the mouse....)

**Comment by mawe42** [comment]
Oh, and maybe we should clarify the documentation for `fluid_settings_dupstr` and `fluid_free`. Remove the "convenience wapper" text and clearly state that using `fluid_free` is required for safe code with mixed static/dynamic C runtimes.

**Comment by jjceresa** [comment]
>and clearly state that using fluid_free is required for safe code with mixed static/dynamic C runtimes.

fluid_free() is required only to ensure that the memory are freed by the same dll that made this memory allocation. This is due to the fact that when a dll x does a memory allocation this memory comes from this dll x's heap. Each distinct dll have a distinct heap and a memory bloc allocated from dll's heap x must be freed to this dll's heap x (not to dll's heap y).

The useful option proposed by Pedro has nothing to do with fluid_free(). In general for Windows applications we must avoid to mix static/dynamic C runtimes dll. The rule is to use dynamic C runtime dll most of the times as possible because the application is using the dll already loaded and used by the OS itself. Also this makes the application code size far less bigger.
Sometimes static C runtime introduces side effect. For example calling the `locale` API doesn't behaves correctly.

**Comment by mawe42** [comment]
> fluid_free() is required only to ensure that the memory are freed by the same dll that made this memory allocation. [...]

Yes... that's what I proposed we should document. Or maybe I didn't understand your point?

> The useful option proposed by Pedro has nothing to do with fluid_free().

Yes, I understand. But the problem with freeing memory allocated by FS without with `free` instead of `fluid_free()` only happens with different static CRTs, does it not?

> In general for Windows applications we must avoid to mix static/dynamic C runtimes dll. [...] Sometimes static C runtime introduces side effect.

That sounds like a good case for introducing the option and having it default to false, in my opinion. And only switch it on for the WinXP CI builds.

**Comment by pedrolcl** [comment]
I don't really understand the hostility towards this proposal, that only adds an option and do not force anybody to use it.

Without this option, it is only slightly inconvenient for me, because I would need to edit the file CMakeLists.txt each time before building Fluidsynth with MSVC. That is only slightly more uncomfortable than building Fluidsynth in Windows, because it is already quite uncomfortable.

> If I understand correctly, the problem exists since 2018 with QSynth using FluidSynth >= 2.0.0rc1 compiled with MSVC

Since I've been the packager of qsynth for windows, we have always used MinGW for building the published binaries, never MSVC. Because the MinGW C runtime does not have this kind of asserts, it does not matter if you call `free` or `fluid_free`. And if you use MSVC and the dynamically linked CRT for all your application modules, it doesn't matter either.

Why then I am interested in this issue, then? because I like to keep all my options open. I try to build and test each new version with both compilers, and compare results. When something ugly is detected, I try to avoid exposing the users to our miseries. I hope you understand my point of view.

**Comment by mawe42** [comment]
I don't think there is or was hostility involved. Quite the contrary: JJC is very in favour, Tom immediately said he was open for this change but simply pointed out that the crash in QSynth should not be the main motivation for the change. And for FWIW... I also think it's a good idea.

**Comment by jjceresa** [comment]
>That sounds like a good case for introducing the option and having it default to false, in my opinion. And only switch it on for the WinXP CI builds.

This would be by far the better choice hopping that isn't to difficult to change the WinXP CI builds configuration.

**Comment by pedrolcl** [comment]
> I don't think there is or was hostility involved. Quite the contrary: JJC is very in favour

Sorry, thanks for your support! I hope that Tom change his mind.

**Comment by derselbst** [comment]
> In general for Windows applications we must avoid to mix static/dynamic C runtimes dll. The rule is to use dynamic C runtime dll most of the times as possible because the application is using the dll already loaded and used by the OS itself.

No offense JJC, but to me, this sounds more like your personal conviction. I am really missing technical arguments that support the _must be done_ arguing.

> Also this makes the application code size far less bigger.

AFAIK, the price for saving a few MB of size is paid by the end-user, who must have installed an adequate version of the MSVC++ Redist package, so that libs like `VCRUNTIME140.dll` will be available, isn't it? I'm afraid WinXP users wouldn't be amused about that.

> But the problem with freeing memory allocated by FS ~without~ with free instead of fluid_free() only happens with different CRTs, does it not?

Technically, freeing memory that was allocated in a "foreign world" is undefined behaviour. One just shouldn't do it. The implementation in use will either make this obvious or silently ignore or fail later on.

> > I don't think there is or was hostility involved. Quite the contrary: JJC is very in favour
> 
> Sorry, thanks for your support! I hope that Tom change his mind.

Like Marcus said, I'm still open, so I haven't really changed my mind.

For me, the question remains: How to implement this? Yes, we could take Pedro's proposed PR and add yet another `enable-something` option. Or, we could use the proper setting provided by cmake: [`CMAKE_MSVC_RUNTIME_LIBRARY`](https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html)

I would vote for the latter approach, because I think that esp. Windows packagers should be more familiar with a cmake option  than with a custom one. However, this would require bumping the minimum cmake version to 3.15. Which I think is ok, because that version was released almost 2 years ago and we cannot stay legacy forever.

Pedro's proposal on the other hand would introduce an "option redundancy" with newer cmake versions. We have a similar redundancy with the `enable-debug` option which I'm not happy with.

Opinions?

**Comment by mawe42** [comment]
Using `CMAKE_MSVC_RUNTIME_LIBRARY` does sound better than adding another option. But I'm not sure the version bump to 3.15 would be worth it. Requiring CMake >= 3.15 would mean Debian Stable (3.13.4), Raspbian (3.13.4) and Ubuntu 18.04 LTS (3.10.2) users would need to get a new cmake from somewhere. Definitely doable (there seems to be a 3.16.3 Debian Stable backport), but judging by how many "How to install FS on Raspberry Pi" tutorials are out there, Debian/Raspbian Stable is used quite a bit for FluidSynth. This version bump would put another stumbling block in those users way and would give them no benefit in return.

**Comment by pedrolcl** [comment]
> In general for Windows applications we must avoid to mix static/dynamic C runtimes dll. The rule is to use dynamic C runtime dll most of the times as possible because the application is using the dll already loaded and used by the OS itself. Also this makes the application code size far less bigger. Sometimes static C runtime introduces side effect. For example calling the locale API doesn't behaves correctly.

I agree. In support of your arguments, I want to share these two documents: [cross crt issues](https://github.com/GNOME/glib/blob/master/README.win32.md#cross-crt-issues) from a document in the Glib repository, which points to [this one](https://docs.microsoft.com/en-us/cpp/c-runtime-library/potential-errors-passing-crt-objects-across-dll-boundaries?view=msvc-160&viewFallbackFrom=vs-2019) by Microsoft. Both documents insist that using different CRT in different modules should be avoided. Fluidsynth uses Glib, among other libraries, and these libs have other dependencies as well. In order to run in XP I suppose that Fluidsynth and its dependencies are all compiled statically, and generate a single big EXE linked to the static CRT, which is acceptable for this concrete case. On the other hand, if every DLL has its own CRT linked, this is precisely what the documents are warning to avoid.

> the price for saving a few MB of size is paid by the end-user, who must have installed an adequate version of the MSVC++ Redist package

One version of the CRT is already included in each Windows version. The users only need to install newer versions if they want to run newer applications, but only once. It is installed system-wide even when it is included by some random MSI of a game or a store application. Windows updates automatically the runtimes when needed, which is not possible for applications statically linked. My windows setup packages always include the latest required runtime redistributables.

> bumping the minimum cmake version to 3.15.

I am all in favor of [modern cmake](https://cliutils.gitlab.io/modern-cmake/), but I would prefer to open another discussion regarding this. I offer my help on this endeavour, but there are more things to consider than the nomenclature of a single variable. For instance: [GnuInstallDirs](https://cmake.org/cmake/help/v3.1/module/GNUInstallDirs.htm) instead of DefaultDirs.cmake

Regarding the spread of CMake versions: https://repology.org/projects/?search=cmake

**Comment by mawe42** [comment]
Thanks for that "Modern CMake" link, very interesting!

One thing that struck me: "CMake 3.14: [...] Windows XP and Vista support has been dropped." (https://cliutils.gitlab.io/modern-cmake/chapters/intro/newcmake.html)

**Comment by derselbst** [comment]
> I am all in favor of modern cmake, but I would prefer to open another discussion regarding this.

I have a few other modernization points on my agenda. I'll start a discussion for both, soon.

> But I'm not sure the version bump to 3.15 would be worth it. Requiring CMake >= 3.15 would mean Debian Stable (3.13.4), Raspbian (3.13.4) and Ubuntu 18.04 LTS (3.10.2) users would need to get a new cmake from somewhere.
> CMake 3.14: [...] Windows XP and Vista support has been dropped

The compromise would be to not bump the minimum version and issue a warning instead. Like:

```cmake
if(POLICY CMP0091)
  cmake_policy(SET CMP0091 NEW)
else()
  message(DEPRECATION "Your version of CMake is old. A future version of fluidsynth may require CMake 3.15 or later.")
endif()
```

People who use cmake >= 3.15 can override `CMAKE_MSVC_RUNTIME_LIBRARY`. All others rely on cmake's default behaviour (i.e. the regex replacement stuff would be removed). Does that sound like a plan?

**Comment by mawe42** [comment]
> The compromise would be to not bump the minimum version and issue a warning instead.

If we can make optional use of the `CMAKE_MSVC_RUNTIME_LIBRARY` feature, then I would vote against a deprecation warning. Or maybe only issue a warning if `CMAKE_MSVC_RUNTIME_LIBRARY` is actually set by the user?

If we really loose XP and Vista support with the switch to 3.15, then that is much more than a simple "deprecation" and would mean we would have to announce that much more visibly, in my opinion.

**Comment by derselbst** [comment]
I'm not quite sure what you mean by optional use. However, using `CMAKE_MSVC_RUNTIME_LIBRARY` requires setting the above policy.

Ok. So no deprecation warning for now, until we've discussed the already mentioned modernization points. I'll open a discussion soon.

**Comment by pedrolcl** [comment]
> People who use cmake >= 3.15 can override CMAKE_MSVC_RUNTIME_LIBRARY. All others rely on cmake's default behaviour (i.e. the regex replacement stuff would be removed). Does that sound like a plan?

If I understood correctly, the (detailed) plan would be:

1. Remove in CMakeLists.txt all the lines from `foreach ( flag_var )` to `endforeach()`, so the variables `CMAKE_<LANG>_FLAGS_<CONFIG>` won't be modified by the script anymore.
2. If the user wants the flags /MT or /MTd when compiling for MSVC, he should set the value of the [CMAKE_<LANG>_FLAGS](https://cmake.org/cmake/help/git-stage/variable/CMAKE_LANG_FLAGS.html#cmake-lang-flags) variable in the command line.
3. Instead of 2. when using CMake>=3.15, users could set the variable `CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` or `CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDebug`for the same results.
4. In absence of one of the above alternatives, the CMake default settings (/MD or /MDd) would be used to compile with MSVC.

Fine for me, of course.

**Comment by derselbst** [comment]
Exactly.

Ok, I'll draft a PR.
